internal/pkg/pluginengine/change.go

```diff
@@ -96,6 +101,14 @@ func execute(smgr statemanager.Manager, changes []*Change) map[string]error {
 		var err error
 		var returnValue map[string]interface{}
 
+		log.Info("tool raw changes: ", c.Tool.Options)
+		// fill ref inputs
+		err = FillInputParams(smgr, c.Tool.Options)
```

Comment:
I think this is probably too late. Maybe we can move the logic into configloader. Update the outputs in-place immediately after the conf is loaded. If there are errors in the outputs, for example, if the key doesn't exist in the outputs at all, an error can be thrown out earlier. We do not have to wait till "change" (apply/delete).

Reply:
This was done before, but there are some issues.

Because the reference ${{}}  depends on the result of the previous plugin.